### ISSUE TITLE
Module not found error 

### ISSUE THREAD

**Issue by sciai-ai** [Masao-Someki/espnet_onnx#29]
Hi, I am getting the following error 

`from espnet_onnx.espnet_onnx.export import TTSModelExport
`

<img width="954" alt="image" src="https://user-images.githubusercontent.com/52277510/173371269-38b68cba-2ebd-4fa9-b0d2-19b7486dd7be.png">

**Comment by Masao-Someki** [comment]
Hi, @sciai-ai, thank you for reporting the bug!!!

Sorry, I found that `__init__.py` is not in the `espnet_onnx.export.layers` directory...
I fixed the bug in [581eb2b](https://github.com/Masao-Someki/espnet_onnx/commit/581eb2b55e3282d57686145bce19ca2d93c111c0), and released the bug-fixed version(0.1.9)
You can use the bug-fixed version by re-installing from pip!

**Comment by Masao-Someki** [comment]
You cannot export mbmelgan. The supported vocoders are listed [here](https://github.com/espnet/espnet/blob/fc62b1ce3e50c5ef8a2ac8cedb0d92ac41df54ca/espnet2/gan_tts/joint/joint_text2wav.py#L46-L51). If you want to convert your custom model, what you have to do is:
1. clone this repository
2. install requirements listed in `requirements_export.txt`
3. create `espnet_onnx.export.tts.models.vocoders.<create your vocoder file>` to define your exportation, such as dummy input and output names
4. modify `espnet_onnx.export.tts.models.__init__.py` to import your custom vocoder

Please check that your model requires only one input. Otherwise, you must implement a new inference class as `espnet_onnx.tts.model.vocoders.vocoder.py`, and modify _build_vocoder method in `espnet_onnx.tts.abs_tts_model.py`

> I also tried with pertained pwgvocoder but get an error that it is not supported

Sorry, pre-trained pwgvocoder is currently not supported. But you can use PWGVocoder trained with the `JointText2Wav` model.

**Comment by sciai-ai** [comment]
> You cannot export mbmelgan. The supported vocoders are listed [here](https://github.com/espnet/espnet/blob/fc62b1ce3e50c5ef8a2ac8cedb0d92ac41df54ca/espnet2/gan_tts/joint/joint_text2wav.py#L46-L51). If you want to convert your custom model, what you have to do is:
> 
> 1. clone this repository
> 2. install requirements listed in `requirements_export.txt`
> 3. create `espnet_onnx.export.tts.models.vocoders.<create your vocoder file>` to define your exportation, such as dummy input and output names
> 4. modify `espnet_onnx.export.tts.models.__init__.py` to import your custom vocoder
> 
> Please check that your model requires only one input. Otherwise, you must implement a new inference class as `espnet_onnx.tts.model.vocoders.vocoder.py`, and modify _build_vocoder method in `espnet_onnx.tts.abs_tts_model.py`
> 

@Masao-Someki Thanks, I will try the steps you mentioned.

> > I also tried with pertained pwgvocoder but get an error that it is not supported
> 
> Sorry, pre-trained pwgvocoder is currently not supported. But you can use PWGVocoder trained with the `JointText2Wav` model.

I also tried with pretained StyleMelgan_v2, but gets  the same error. Are all the supported vocoders trained with joint text2wav?

**Comment by Masao-Someki** [comment]
@sciai-ai
> I also tried with pretained StyleMelgan_v2, but gets the same error. Are all the supported vocoders trained with joint text2wav?

Currently, yes. However, you can create a new JointText2Wav instance for exportation. You can set pretrained tacotron2 as text2mel and pretrained pwgvocoder as a vocoder. I've never tested this before, but I think you can export and use it for inference in this way.

> Also it it possible to use onnx text2mel,such as tacotron2.onnx model with a non-onxx exported vocoder, such as, pre-trained pwgvocoder. 

Yes. Please check that the `vocoder_type` in config.yaml is set to `not_used` so that the espnet_onnx will not build vocoder.
config.yaml is located in `~/.cache/espnet_onnx/<your model>/config.yaml`.

```
model = Text2Speech('your model')

# inference only text2mel
mel = model(text)['feat_gen']

wav = your_vocoder(mel)
```
Note that the type of `mel` is `np.ndarray`, so you have to convert it to torch.Tensor.

**Comment by sciai-ai** [comment]
@Masao-Someki 
> > Also it it possible to use onnx text2mel,such as tacotron2.onnx model with a non-onxx exported vocoder, such as, pre-trained pwgvocoder.
> 
> Yes. Please check that the `vocoder_type` in config.yaml is set to `not_used` so that the espnet_onnx will not build vocoder. config.yaml is located in `~/.cache/espnet_onnx/<your model>/config.yaml`.
> 
> ```
> model = Text2Speech('your model')
> 
> # inference only text2mel
> mel = model(text)['feat_gen']
> 
> wav = your_vocoder(mel)
> ```
> 
> Note that the type of `mel` is `np.ndarray`, so you have to convert it to torch.Tensor.

I tried this approach, I noticed a few issues 

1. Text2Speech.from_pretrained ignores my arguments, such as min len ration, attn constraint. I have used these parameters in default espnet and they impact the output. Can you please check if the parameters are not ignored in your code?
`Text2Speech.from_pretrained(model_file=model_path,vocoder_file=vocoder_path,
                                     train_config=train_config,vocoder_config=vocoder_config,
                                     device=device_type,
    # Only for Tacotron 2
    threshold=0.5,
    minlenratio=0.0,
    maxlenratio=20.0,
    use_att_constraint=True,
    backward_window=3,
    forward_window=7,
    # Only for FastSpeech & FastSpeech2
    speed_control_alpha=1.0,)`

2. I need to use  `normalize_before=True`   in `wav = vocoder.inference(torch.tensor(mel), normalize_before=True)`, my vocoder and text2mel are trained on the same data and in default epsnet I never use this noramlization. I wonder why I need to do this in espent_onnx.

4. I tried with GPU onnx and get the error while exporting

<img width="1251" alt="image" src="https://user-images.githubusercontent.com/52277510/173731319-f718f478-e3e0-44ed-9ad3-7344b26a0029.png">

**Comment by Masao-Someki** [comment]
> Text2Speech.from_pretrained ignores my arguments,i.e, forward and backward window, attn constraint. I have used these parameters in default espnet and they impact the output. Can you please check if the parameters are not ignored in your code?
Text2Speech.from_pretrained(model_file=model_path,vocoder_file=vocoder_path, train_config=train_config,vocoder_config=vocoder_config, device=device_type, threshold=0.5, minlenratio=0.0, maxlenratio=20.0, use_att_constraint=True, backward_window=3, forward_window=7)

Sorry, I commented out the script related to `use_att_constraint` when debugging, and I completely forgot to check and implement those...
I will implement and fix this in the next PR!

> I need to use normalize_before=True in wav = vocoder.inference(torch.tensor(mel), normalize_before=True), my vocoder and text2mel are trained on the same data and in default epsnet I never use this noramlization. I wonder why I need to do this in espent_onnx.

I think this is because the normalization is applied after text2mel inference. Please check that `normalize.use_normalize` in the config.yml is set to false. If this is true, then normalization will be applied after text2mel inference.

> I tried with GPU onnx and get the error while exporting

Your model is on the GPU. Please check that all parameters are on the CPU. 

> Decoding using text2speech -> feat_gen -> vocoder produces different results than text2speech -> wav

I think this is related to the second question. Would you check the normalization config and try this again?

**Comment by sciai-ai** [comment]
> > I tried with GPU onnx and get the error while exporting
> 
> Your model is on the GPU. Please check that all parameters are on the CPU.
> 
> >

Can you advice me how to check this, The only parameter I changed was `device_type='cuda'` in Text2Speech  but still get the error

**Comment by Masao-Someki** [comment]
When you try to export your model, you have to set `device_type='cpu'`.
You can use your GPU with `onnxruntime-gpu`. Please refer to [How to use GPU on espnet_onnx](https://github.com/Masao-Someki/espnet_onnx#how-to-use-gpu-on-espnet_onnx)

**Comment by sciai-ai** [comment]
Sorry i didn't understand, i use gpu because i want to do gpu inference with onnxruntim_gpu installed.

**Comment by Masao-Someki** [comment]
Could you provide a snippet for reproduction?
I don't get the same problem, so I want to see your code..

**Comment by sciai-ai** [comment]
```
from espnet_onnx import Text2Speech
from parallel_wavegan.utils import load_model
from IPython.display import display, Audio
import torch 
import time

vocoder = load_model(vocoder_path).to('cpu').eval()
# vocoder.remove_weight_norm()

# text2speech2 = Text2Speech.from_pretrained(model_file=model_path,
#                                      train_config=train_config,
#                                      device='cuda',
#     # Only for Tacotron 2
#     threshold=0.5,
#     minlenratio=0.0,
#     maxlenratio=20.0,
#     use_att_constraint=True,
#     backward_window=3,
#     forward_window=7,
#     # Only for FastSpeech & FastSpeech2
#     speed_control_alpha=1.0,)



PROVIDERS = ['CUDAExecutionProvider']

tag_name = 'taco2'
start = time.time()

with torch.no_grad():
    torch.manual_seed(3)
    text2speech = Text2Speech(tag_name, use_quantized=False, providers=PROVIDERS)
    mel = text2speech(text)['feat_gen']
    torch.manual_seed(343)
    wav = vocoder.inference(torch.tensor(mel), normalize_before=True)
    
wav = wav.view(-1).detach().numpy()
rtf = (time.time() - start) / (len(wav) / 22050)
# print(f"RTF = {rtf:5f}")
elapsed_time = time.time() - start
print(elapsed_time)

display(Audio(wav, rate=22050))
```

**Comment by sciai-ai** [comment]
> I found some problems in your code, so please check the following:
> 
> 1. Your test code includes loading time. To check the inference time, you have to instantiate `text2speech` before `start=time.time()` line.
I changed this and now the gput-onnx time is ~4.73 seconds, still much slower than gpu inference.

> 2. `espnet_onnx.Text2Speech` does not use PyTorch in inference. I think `torch.manual_seed(3)` is not required.
> 3. Please check your config file. If `normalize.use_normalize` is false, then I think you don't need to set `normalize_before=True` in vocoder.inference
This is ok because if the normalization is not set properly the synthesiexed speech quality is very bad. In mycase it sounds good. Here is config file 

<img width="397" alt="image" src="https://user-images.githubusercontent.com/52277510/173874839-6e045de1-b731-4989-9092-eff968bf23a2.png">

**Comment by Masao-Someki** [comment]
I'm not sure in detail, but the first inference of the onnx model after loading the model takes quite a long time.
I think you can get faster performance after the first inference.

```
text2speech = Text2Speech(tag_name)
text = 'hello world'

out = text2speech(text) # <-- slow
out = text2speech(text) # <-- fast
out = text2speech(text) # <-- fast
```

I got the long silence problem with the quantized version...
I will investigate this, but please let me know if you found some weird thing..

**Comment by sciai-ai** [comment]
Let's focus on the non quantized version first, I tried many inferences, its still much slower than normal gpu inference. I think onnx gpu should be much faster than gpu inference?

**Comment by Masao-Someki** [comment]
@sciai-ai
I tested with torch and onnx with GPU. As you said, I found that the execution time for the onnx version is slower than the torch version...
The length of the text is 175.

- torch GPU : 0.62479 sec
- onnx GPU : 0.86348 sec

I guess that the number of Conv module would do something with this problem, but I have to investigate more about this problem..

**Comment by sciai-ai** [comment]
Hmm, do you observe the same problem for fastspeech2 as well?

How about converting onnx/pytorch to tensorrt?

**Comment by Masao-Someki** [comment]
I tested with fastspeech2 (`text='hello world'`) and I found that the performance was improved.
- torch: 0.08298 sec
- onnx: 0.03790 sec

I've never convert models to tensorrt, so I'm not sure about this...

**Comment by sciai-ai** [comment]
is this based on GPU?

**Comment by Masao-Someki** [comment]
I tested again and got the following result.
- torch: 0.00797 sec
- onnx: 0.037399 sec

It seems that the depthwise convolution is slow in fp32 on GPU. This might be the reason for this bad performance...

**Comment by sciai-ai** [comment]
Is this fasrspeech2? Torch is 10 times faster than your last result?

**Comment by Masao-Someki** [comment]
The performance of onnx tts depends on what module you used in your model.
I've never experienced this kind of performance degradation with ASR models, so I think the cause is related to the number of conv modules..

**Comment by sciai-ai** [comment]
How about the performance for VITS?
